### PR TITLE
resolved: actually check authenticated flag of SOA transaction

### LINKED ISSUE #25676: resolved DNSSEC validation can be bypassed by MITM
### systemd version the issue has been seen with

251.8-1-manjaro

### Used distribution

Manjaro

### Linux kernel version used

5.15.81-1-MANJARO

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

When `DNSSEC=yes` is set and querying a DNSSEC-enabled domain, systemd-resolved should only accept records that are signed correctly.

### Unexpected behaviour you saw

systemd-resolved accepts records of DNSSEC-signed domains even when they have no signature, allowing man-in-the-middles (or the upstream DNS resolver) to manipulate records.

Please note that I have repeatedly tried to report this issue to the systemd-security mailing list. I sent a mail with the details and reproduction steps to the systemd-security address on Oct 6th, Oct 14th and Oct 22nd. I first received a reply on Oct 24th from Lennart Poettering, asking for some basic system information, which I provided on the same day. Then I received no further reply, so I sent the report one more time to systemd-security on Nov 7th. The next day Poettering replied to my previous mail, asking for a debug log, but not directly acknowledging the issue. I sent debug logs on Nov 10th and haven't heard back from anyone since then.

### Steps to reproduce the problem

The issue can be reproduced with an unbound server and a typetransparent local-zone to simulate an attacker modifying select records. `cloudflare.net` is used as an example domain since it is [signed and valid](https://dnssec-analyzer.verisignlabs.com/cloudflare.net). In this example, the DNS server is run on `192.168.1.1` and the systemd client has any address in the same /24 network.

unbound.conf:
```
server:
  # Allow running unbound unprivileged
  username: ""
  use-syslog: no
  chroot: ""
  pidfile: ""

  interface: 0.0.0.0
  port: 10053
  access-control: 192.168.1.1/24 allow

  local-zone: www.cloudflare.net typetransparent
  local-data: "www.cloudflare.net. 3600 IN A 1.2.3.4"

forward-zone:
  name: .
  forward-addr: 1.1.1.1
  forward-addr: 1.0.0.1
  forward-first: no
```

1. On the "attacker" machine, install unbound and run `unbound -d -v -c unbound.conf`. 
2. On the machine running systemd-resolved, run `sudo systemd-resolve --set-dns 192.168.1.1:10053 --set-dnssec yes --interface enp0s3`. (Replace IP and interface name accordingly).
3. Run `resolvectl status` to confirm our DNS server is used and output contains `DNSSEC=yes/supported`.
4. Run `resolvectl query cloudflare.net` which should produce the correct reply and `Data is authenticated: yes`, confirming that the domain is DNSSEC-signed.
5. Run `resolvectl query www.cloudflare.net`:
```
$ resolvectl query www.cloudflare.net
www.cloudflare.net: 1.2.3.4                    -- link: enp0s3

-- Information acquired via protocol DNS in 25.6ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```
systemd-resolved accepts the bogus response. Running `dig @127.0.0.53 www.cloudflare.net` also shows the status is `NOERROR`. When resolved is configured as system resolver (e.g. through `/etc/resolv.conf`), `curl -v http://www.cloudflare.net` tries to connect to `1.2.3.4`, showing that applications are successfully tricked with the manipulated record.

Note that resolved does state the data is unauthenticated and also does not set the AD bit when queried on `127.0.0.53`.

Repeat the test but replace `typetransparent` with `static` in the unbound configuration. In this case, resolved rejects the response as expected:
```
$ resolvectl query www.cloudflare.net
www.cloudflare.net: resolve call failed: DNSSEC validation failed: no-signature
```
(Note that due to another bug https://github.com/systemd/systemd/issues/24827, resolved will still reply with the records on `127.0.0.53`, but it has `status: SERVFAIL`, so applications generally won't use this answer)

The difference is that `static` replaces all data for the `www.cloudflare.net` domain, whereas `typetransparent` only replaces the A record(s). Both strip the RRSIG records from affected queries.

My knowledge on DNS isn't great, but as far as I know no unsigned record should be allowed anywhere under the zone cloudflare.net, unless there was an insecure delegation (NS record without corresponding DS record).
However, it appears only signatures of some types of records actually matter to resolved, for example the SOA record. The RRSIG signatures on A/AAAA (and probably more) records can simply be stripped and resolved will accept the data without ensuring its validity.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
systemd-resolved debug logs, including all output after running `resolvectl query www.cloudflare.net` are posted in this gist, separate for the typetransparent and the static test case: https://gist.github.com/AgentOak/11c64ff8701c4cd853d6b99382a40704
```

### PR DESCRIPTION
**Pull request by msekletar**
Fixes #25676

### CODE REVIEW

**Comment by msekletar** [comment]
I tried to take a stab at #25676 and this is what I came up with,  it seems to fix the issue but please review it carefully as I am not really resolved expert.

File: src/resolve/resolved-dns-transaction.c [hunk 1]
```diff
@@ -2835,7 +2835,7 @@ static int dns_transaction_requires_rrsig(DnsTransaction *t, DnsResourceRecord *
                         /* We found the transaction that was supposed to find the SOA RR for us. It was
                          * successful, but found no RR for us. This means we are not at a zone cut. In this
                          * case, we require authentication if the SOA lookup was authenticated too. */
-                        return FLAGS_SET(t->answer_query_flags, SD_RESOLVED_AUTHENTICATED);
+                        return FLAGS_SET(dt->answer_query_flags, SD_RESOLVED_AUTHENTICATED);
```
Comment (yuwata) [inline src/resolve/resolved-dns-transaction.c:2838]: Exactly same mistake can be found in the above case block. Could you also fix it?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/resolve/resolved-dns-transaction.c (hunk 1 above)
Comment (msekletar) [reply src/resolve/resolved-dns-transaction.c:2838]: Done

**Review by yuwata — APPROVED** [review]

**Comment by yuwata** [comment]
@poettering Could you double check?

**Comment by poettering** [comment]
ouch.

lgtm

**Comment by pemensik** [comment]
I have taken a look into this commit and it seemed a *way too* simple to fix the issue properly. I made my own rawhide build with added only this patch [on top](https://src.fedoraproject.org/fork/pemensik/rpms/systemd/tree/CVE-2023-7008) of last build.

As I have suspected this changes *IS NOT* sufficient to prevent the cache poisoning. I have used simple addition to libvirt's dnsmasq backed network.

```
# sudo virsh net-edit default
<network>
  <!-- ... -->
  <dns>
    <txt name='example.net' value='v=spf1 +all'/>
    <host ip='127.0.0.255'>
      <hostname>example.net</hostname>
    </host>
  </dns>
  <!-- ... -->
</network>
```

Unless I have made mistake with build, this is not sufficient fix. It may protect SOA query itself, but does not protect other records. It is still possible to spoof own records. All example domains ARE signed, so any their spoofing should make them broken and inaccessible.

```
# resolvectl dnssec
Global: yes
Link 2 (host0): yes
# rpm -q systemd-resolved
systemd-resolved-255.1-2.fc40.x86_64
# resolvectl query example.net
example.net: 127.0.0.255                       -- link: host0
             2606:2800:220:1:248:1893:25c8:1946 -- link: host0

-- Information acquired via protocol DNS in 5.9ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: cache network
# resolvectl query -t txt example.net
example.net IN TXT "v=spf1 +all"                            -- link: host0

-- Information acquired via protocol DNS in 2.1ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
# delv example.net
;; broken trust chain resolving 'example.net/A/IN': 127.0.0.53#53
;; resolution failed: broken trust chain
# unbound-host -rvD example.net
example.net has address 127.0.0.255 (BOGUS (security failure))
validation failure <example.net. A IN>: SERVFAIL no DS for DS net. while building chain of trust
example.net has IPv6 address 2606:2800:220:1:248:1893:25c8:1946 (BOGUS (security failure))
validation failure <example.net. AAAA IN>: key for validation net. is marked as invalid because of a previous no DS
example.net mail is handled by 0 . (BOGUS (security failure))
validation failure <example.net. MX IN>: key for validation net. is marked as invalid because of a previous no DS
```

**Review by pemensik — COMMENTED** [review]
I am afraid it is not as simple as demanding AD bit in SOA queries. First problem is those SOA queries are not necessary anyway. But it is possible to keep SOA query intact and replace just record type you want to modify. That is done by *dnsmasq* for example, it does not care about zone definitions.

**Comment by pemensik** [comment]
False alarm, this seems to fix the issue well, but Fedora package does not contain instruction to restart *systemd-resolved.service*, therefore it had no immediate effect.

**Pull request merged**